### PR TITLE
Throw Winml_err_invalid_binding if binding gpu resource on cpu device

### PR DESCRIPTION
**Pull request by ryanlai2**
Get the ProviderType from the execution provider and check if the provider is DmlExecutionProvider. This is done through the WinMLAdapter.

If the provider type isn't DmlExecutionProvider, then throw WINML_ERR_INVALID_BINDING

Here is where the checked used to be in the unlayered version of WinML.
https://github.com/microsoft/onnxruntime/blob/5780b864a15513fda4eadbfc2b5345fefe70b5ec/winml/lib/Api.Core/inc/TensorBaseHelpers.h#L16

### CODE REVIEW

File: winml/adapter/WinMLAdapter.cpp [hunk 1]
```diff
@@ -600,6 +600,11 @@ class WinMLAdapter : public Microsoft::WRL::RuntimeClass<
   }
   WINML_CATCH_ALL_COM
 
+  const char* STDMETHODCALLTYPE GetProviderType(
+      onnxruntime::IExecutionProvider* provider) noexcept override {
```
Comment (tiagoshibata) [inline winml/adapter/WinMLAdapter.cpp]: Can it be a const pointer?

File: winml/lib/Api/impl/TensorBase.h [hunk 2]
```diff
@@ -104,7 +104,9 @@ struct TensorBase : TBase {
     auto session_impl = context.session.as<winrt::Windows::AI::MachineLearning::implementation::LearningModelSession>();
     auto provider = session_impl->GetExecutionProvider();
     WINML_THROW_IF_FAILED(adapter_->GetProviderMemoryInfo(provider, dml_memory.put()));
-
+    WINML_THROW_HR_IF_TRUE_MSG(WINML_ERR_INVALID_BINDING,
+                               "DmlExecutionProvider" != adapter_->GetProviderType(provider),
```
Comment (walrusmcd) [inline winml/lib/Api/impl/TensorBase.h]: >DmlExecutionProvider [](start = 32, length = 20)

onnxruntime::kDmlExecutionProvider

Comment (walrusmcd) [reply winml/lib/Api/impl/TensorBase.h]: actually, the code already in TensorBase does this pattern to detect CPU v GPU.   copy that?

    auto spSession = context.session.as<winrt::Windows::AI::MachineLearning::implementation::LearningModelSession>();
    auto spDevice = spSession->Device().as<winrt::Windows::AI::MachineLearning::implementation::LearningModelDevice>();
    if (spDevice->IsCpuDevice()) {


look at line # 214

---
In reply to: [355645725](https://github.com/microsoft/onnxruntime/pull/2589#discussion_r355645725) [](ancestors = 355645725)

File: winml/adapter/WinMLAdapter.cpp (hunk 1 above)
Comment (ryanlai2) [reply winml/adapter/WinMLAdapter.cpp]: yep yep!

File: winml/lib/Api/impl/TensorBase.h (hunk 2 above)
Comment (walrusmcd) [reply winml/lib/Api/impl/TensorBase.h]: i see what you did, you actually went to the old code this was in and just brought it fwd.  that makes sense :) 

---
In reply to: [355646168](https://github.com/microsoft/onnxruntime/pull/2589#discussion_r355646168) [](ancestors = 355646168,355645725)

Comment (ryanlai2) [reply winml/lib/Api/impl/TensorBase.h]: Good catch! I wonder why we got it from the provider previously.

**Pull request merged**